R/model-diff.R

```diff
@@ -0,0 +1,138 @@
+#'   See "`.file` argument" in Details section.
+#' @param ... arguments passed through to methods. (Currently none.)
+#' @param .viewer If `FALSE`, the default, prints diff to console or renders in
+#'   Rmd. If `TRUE`, render the diff in the Viewer window. Note: this option
+#'   will hold the console with `Press ENTER to continue...` so it is only
+#'   recommended for interactive use.
+#' @export
+model_diff <- function(.mod, .mod2 = NULL, .file = "model", ..., .viewer = FALSE) {
+  UseMethod("model_diff")
+}
+
+#' @rdname model_diff
+#' @export
+model_diff.bbi_nonmem_model <- function(
+  .mod,
+  .mod2 = NULL,
+  .file = c("model"),
+  ...,
+  .viewer = FALSE
+) {
+  .file <- match.arg(.file)
+
+  .mod2 <- model_diff_get_comp(.mod, .mod2)
+
+  .file1 <- get_model_path(.mod)
+  .file2 <- get_model_path(.mod2)
+
+  model_diff_impl(.file1, .file2, .viewer = .viewer)
+}
+
+
+###################################
+# PRIVATE IMPLEMENTATION FUNCTIONS
+###################################
+
+#' Private helper to get a valid comparison model
+#' @inheritParams model_diff
+#' @keywords internal
+model_diff_get_comp <- function(.mod, .mod2) {
+  check_yaml_in_sync(.mod)
+
+  if (is.null(.mod2)) {
+    based_on_path <- get_based_on(.mod)
+    mod_id <- get_model_id(.mod)
+    if (length(based_on_path) == 0) {
+      stop(paste(glue("Model {mod_id} has no models in `based_on`."), MODEL_DIFF_ERR_MSG), call. = F)
+    } else if (length(based_on_path) > 1) {
+      stop(paste(glue("Model {mod_id} has multiple models in `based_on`."), MODEL_DIFF_ERR_MSG), call. = F)
+    } else {
+      .mod2 <- read_model(based_on_path)
+    }
+  }
+  check_yaml_in_sync(.mod2)
+  return(.mod2)
+}
+
+#' Private helper to diff two model files
+#' @importFrom diffobj diffFile
+#' @param model_A Path to first model file
+#' @param model_B Path to second model file
+#' @inheritParams model_diff
+#' @keywords internal
+model_diff_impl <- function(model_A, model_B, .viewer) {
+  # check inputs
+  if (!all(c(
+    checkmate::check_string(model_A),
+    checkmate::check_string(model_B),
+    fs::file_exists(c(model_A, model_B))
+  ))) {
+    dev_error(paste(
+      "model_diff_impl() got the following:",
+      paste(model_A, collapse =" -- "),
+      paste(model_B, collapse =" -- ")
+    ))
+  }
+
+  suppressSpecificWarning({
+    # if knitting and `results = "asis"` set, render HTML
+    if (isTRUE(getOption('knitr.in.progress')) &&
+        knitr::opts_current$get("results") == 'asis' &&
+        knitr::is_html_output()
+        ) {
+      diffobj::diffFile(
+        model_A,
+        model_B,
+        mode = "sidebyside",
+        format = "html",
+        style = list(html.output = "diff.w.style"),
+        interactive = F
+      )
+    } else {
+      # otherwise, print to either console or viewer
+      diffobj::diffFile(
+        model_A,
+        model_B,
+        mode = "sidebyside",
+        interactive = .viewer
```

Comment:
do we really want it to silently fail if the user does `.viewer=T` but doesn't include `results='asis'`? Should we print a warning to the output like "to include output in knitted file, set chunk option to results=asis"?

Reply:
good call. I'll add a warning if someone is knitting and has `.viewer=T` in there

Reply:
Check it:
<img width="849" alt="Screen Shot 2021-03-03 at 3 28 57 PM" src="https://user-images.githubusercontent.com/7806083/109868072-3937c800-7c35-11eb-9a5c-bfa56a2a68f9.png">

Reply:
I also added a warning for trying to use this in non-HTML knitting. See the bottom of this comment https://github.com/metrumresearchgroup/bbr/pull/342#issuecomment-789150414